WIP: Improve README and main entrypoints into the docs (#11558)
crate/crate: docs/general/ddl/data-types.rst

For example::

  cr> CREATE TABLE my_table (
  ...     bit_mask BIT(4)
  ... );
  CREATE OK, 1 row affected (... sec)

::

  cr> INSERT INTO my_table (
  ...     bit_mask
  ... ) VALUES (
  ...     B'0110'
  ... );
  INSERT OK, 1 row affected  (... sec)

.. HIDE:

    cr> REFRESH TABLE my_table;
    REFRESH OK, 1 row affected (... sec)

::

    cr> SELECT bit_mask FROM my_table;
    +----------+
    | bit_mask |
    +----------+
    | B'0110'  |
    +----------+
    SELECT 1 row in set (... sec)

Inserting values that are either too short or too long results in an error::

  cr> INSERT INTO my_table (
  ...     bit_mask
  ... ) VALUES (
  ...    B'00101'
  ... );
  SQLParseException[bit string length 5 does not match type bit(4)]

.. HIDE:

    cr> DROP TABLE my_table;
    DROP OK, 1 row affected (... sec)

.. _data-types-numeric:

Numeric data
------------

CrateDB supports the following numeric types:

.. contents::
   :local:
   :depth: 1

.. _data-types-floating-point:

.. NOTE::

    The :ref:`REAL <type-real>` and :ref:`DOUBLE PRECISION
    <type-double-precision>` data types are inexact, variable-precision

>     floating-poin t types, meaning that these types are stored as an

Reviewer: ```suggestion
    floating-point types, meaning that these types are stored as an
```

    to ``33658-09-27T01:46:39.999Z``) rounding issues may occur.

A ``TIMESTAMP`` can be further defined as:

.. contents::
   :local:
   :depth: 1

.. _type-timestamp-with-tz:

``WITH TIME ZONE``
..................

If you define a timestamp as ``TIMESTAMP WITH TIME ZONE``, CrateDB will convert
string literals to `Coordinated Universal Time`_ (UTC) using the ``offset``
value (e.g., ``+01:00`` for plus one hour or ``Z`` for UTC).

Example::

    cr> CREATE TABLE my_table (
    ...     ts_tz_1 TIMESTAMP WITH TIME ZONE,
    ...     ts_tz_2 TIMESTAMP WITH TIME ZONE
    ... );
    CREATE OK, 1 row affected (... sec)

::

    cr> INSERT INTO my_table (
    ...     ts_tz_1,
    ...     ts_tz_2
    ... ) VALUES (
    ...     '1970-01-02T00:00:00',
    ...     '1970-01-02T00:00:00+01:00'
    ... );
    CREATE OK, 1 row affected (... sec)

.. HIDE:

    cr> REFRESH TABLE my_table;
    REFRESH OK, 1 row affected (... sec)

::

    cr> SELECT
    ...     ts_tz_1,
    ...     ts_tz_2
    ... FROM my_table;
    +----------+----------+
    |  ts_tz_1 |  ts_tz_2 |
    +----------+----------+
    | 86400000 | 82800000 |
    +----------+----------+
    SELECT 1 row in set (... sec)

> You can use :ref:`data_format() <scalar-date_format>` to to make the output

Reviewer: ```suggestion
You can use :ref:`data_format() <scalar-date_format>` to make the output
```

value::

    cr> SELECT INTERVAL '1-2 3 4:05:06' YEAR TO MONTH AS result;
    +------------------------+
    | result                 |
    +------------------------+
    | 1 year 2 mons 00:00:00 |
    +------------------------+
    SELECT 1 row in set (... sec)

You can use ``DAY TO HOUR``, as another example, to limit a day-time interval
to days and hours::

    cr> SELECT INTERVAL '3 4:05:06' DAY TO HOUR AS result;
    +-----------------+
    | result          |
    +-----------------+
    | 3 days 04:00:00 |
    +-----------------+
    SELECT 1 row in set (... sec)

.. TIP::

    You can use intervals in combination with :ref:`CURRENT_TIMESTAMP
    <current_timestamp>` to calculate values that are offset relative to the
    current date and time.

    For example, to calculate a timestamp corresponding to exactly one day ago,
    use::

        cr>  SELECT CURRENT_TIMESTAMP - INTERVAL '1' DAY AS result;
        +---------------+
        | result        |
        +---------------+
        | 1623847247282 |
        +---------------+
        SELECT 1 row in set (... sec)

.. _data-types-ip-addresses:

IP addresses
------------

.. _type-ip:

``IP``
''''''

> An ``IP`` is a string representation of an `IP adddress`_ (IPv4 or IPv6).

Reviewer: ```suggestion
An ``IP`` is a string representation of an `IP address`_ (IPv4 or IPv6).
```

    ... );

.. HIDE:

    cr> REFRESH TABLE my_table;
    REFRESH OK, 1 row affected (... sec)

::

    cr> SELECT
    ...     protagonist['name'] as name,
    ...     protagonist['chapter'] as chapter,
    ...     pg_typeof(protagonist['size']) as size
    ... FROM my_table
    ... ORDER BY protagonist['chapter'] ASC;
    +-------------------------------+
    | payload                       |
    +-------------------------------+
    | {"tag": "AT", "value": 30}    |
    | {"tag": "AT", "value": "str"} |
    +-------------------------------+
    SELECT 2 rows in set (... sec)

::

    cr> CREATE TABLE my_table (
    ...     TODO TEXT
    ... );
    CREATE OK, 1 row affected (... sec)

TODO::

    cr> INSERT INTO my_table (
    ...     TODO
    ... ) VALUES (
    ...     'TODO'
    ... );
    CREATE OK, 1 row affected (... sec)

.. HIDE:

    cr> REFRESH TABLE my_table;
    REFRESH OK, 1 row affected (... sec)

TODO::

    cr> SELECT * FROM my_table;
    +--------------+
    | TODO         |
    +--------------+
    | TODO         |
    +--------------+
    SELECT 1 row in set (... sec)

.. HIDE:

    cr> DROP TABLE my_table;
    DROP OK, 1 row affected (... sec)

.. NOTE::

>     Given that dynamically added sub-columns of an ``IGNORED`` objects are not

Reviewer: ```suggestion
    Given that dynamically added sub-columns of an ``IGNORED`` object are not
```

    +----------------------------------+
    SELECT 1 row in set (... sec)

``null`` elements are interpreted as ``null`` (none, absent), if you want the
literal ``null`` string, it has to be enclosed in double quotes.

This variant primarily exists for compatibility with PostgreSQL. The array
constructor syntax explained further above is the preferred way to define
constant array values.

.. _data-types-geo:

Geographic types
================

:ref:`Geographic types <data-types-geo>` are types with :ref:`nonscalar
<gloss-nonscalar>` values representing points or shapes in a 2D world:

.. contents::
   :local:
   :depth: 3

.. _data-types-geo-point:

Geometric points
----------------

.. _type-geo_point:

``GEO_POINT``
'''''''''''''

A ``GEO_POINT`` is a :ref:`geographic data type <data-types-geo>` used to store
latitude and longitude coordinates.

Columns with the ``GEO_POINT`` type are represented and inserted using an array
of doubles in the following format::

    [<lon_value>, <lat_value>]

Alternatively a `WKT`_ string can also be used to declare geo points::

    'POINT ( <lon_value> <lat_value> )'

.. NOTE::

    Empty geo points are not supported.

>     Additionally, if a column is dynamically created the type detection won't

Reviewer: ```suggestion
    Additionally, if a column is dynamically created, the type detection won't
```
